File: sn/handler.go
package sn

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type UserHandler interface {
	WithTx(tx *gorm.DB) UserHandler

	// New create new user, by default no user group will be attached.
	//
	// When avatar is nil, use default avatar.
	New(username string, password string, avatar string) (*User, error)

	// CheckPass check whether user and pass match.
	//
	// If error, return nil,-1,err.
	//
	// If user not found, return nil,1,nil.
	//
	// If pass not match, return nil,2,nil.
	//
	// Return user,0,nil if all match.
	CheckPass(user string, pass string) (*User, int, error)

	// GetAll get all user by condition.
	GetAll(cond *Condition) ([]*User, error)

	// Get get user by id.
	Get(id uuid.UUID) (*User, error)

	// GetByName get user by name.
	//
	// Return nil,nil when user not found.
	GetByName(name string) (*User, error)

	// Count count user by condition.
	Count(cond *Condition) (int64, error)

	// Kick kick user id login.
	Kick(id uuid.UUID) error

	// Reset reset user password by id, return new password.
	Reset(id uuid.UUID) (string, error)

	// Update update user infos by user.ID.
	// Password will be hashed after the function call.
	// Users will be kicked if password is updated.
	Update(column []string, user *User) error

	// Delete delete the entire user, including link and permission.
	Delete(id uuid.UUID) error
}

type GroupHandler interface {
	WithTx(tx *gorm.DB) GroupHandler

	// New create new usergroup.
	New(name string, note string) (*Group, error)

	// Link link all uid user to all gid group.
	Link(uid []uuid.UUID, gid []uuid.UUID) ([]*UserGroupLink, error)

	// Unlink unlink user and group.
	//
	// If uid!=nil and gid!=nil, remove each uid with each gid.
	// If uid==nil, remove all users in each gid.
	// If gid==nil, remove all groups in each uid.
	Unlink(uid []uuid.UUID, gid []uuid.UUID) error

	// GetGroupAllUser get group id all users.
	GetGroupAllUser(id uuid.UUID, cond *Condition) ([]*UserGroupLink, error)

	// CountGroupAllUser count group user by condition.
	CountGroupAllUser(id uuid.UUID, cond *Condition) (int64, error)

	// GetAll get all group by condition.
	GetAll(cond *Condition) ([]*Group, error)

	// Get get group by id.
	Get(id uuid.UUID) (*Group, error)

	// GetByName get group by name.
	GetByName(name string) (*Group, error)

	// GetUserAllGroup get user id all groups.
	GetUserAllGroup(id uuid.UUID) ([]*UserGroupLink, error)

	// Count count group by condition.
	Count(cond *Condition) (int64, error)

	// Update update group infos by group.ID.
	Update(column []string, group *Group) error

	// Delete delete the entire group, including link and permission.
	Delete(id uuid.UUID) error
}

type PermissionHandler interface {
	WithTx(tx *gorm.DB) PermissionHandler

	// AddToUser add permission to user uid.
	AddToUser(uid uuid.UUID, perm []*PermEntry) (ret []*PermissionLink, err error)

	// AddToGroup add permission to group gid.
	AddToGroup(gid uuid.UUID, perm []*PermEntry) (ret []*PermissionLink, err error)

	// GetEntry return permission entries.
	GetEntry() ([]*Permission, error)

	// Grant grant uid or gid with pid and perm.
	//
	// uid and gid should not be both uuid.Nil, otherwise, nil will be returned.
	// If uid and gid both have value, pid will be granted to all of them.
	//
	// If perm==0, permission will be explicitly forbidden.
	// If perm==-1, permission will be revoked.
	Grant(uid uuid.UUID, gid uuid.UUID, pid uuid.UUID, perm UserPerm) error

	// GetAll find all permission by condition, if join is true,
	// return records will join PermissionList.
	//
	// uid and gid should not be both uuid.Nil or both have value, otherwise,
	// nil,nil will be returned
	GetAll(uid uuid.UUID, gid uuid.UUID, joinUser bool, joinGroup bool, joinPerm bool) ([]*PermissionLink, error)

	// GetUser get user perm list.
	GetUser(uid uuid.UUID) (map[uuid.UUID]*PermEntry, error)

	// GetGroup get group perm list.
	GetGroup(gid uuid.UUID) (map[uuid.UUID]*PermEntry, error)

	// GetUserMerged returns merged user permission list.
	GetUserMerged(uid uuid.UUID) (map[uuid.UUID]*PermEntry, error)

	// Delete delete by permission link id.
	Delete(id uuid.UUID) error

	// DeleteUser delete user uid permission pid.
	//
	// If pid == uuid.Nil, all user permission will be deleted.
	DeleteUser(uid uuid.UUID, pid uuid.UUID) (int64, error)

	// DeleteGroup delete group gid permission pid.
	//
	// If pid == uuid.Nil, all group permission will be deleted.
	DeleteGroup(gid uuid.UUID, pid uuid.UUID) (int64, error)
}

type NotificationHandler interface {
	WithTx(tx *gorm.DB) NotificationHandler
	SetUnread(num int64)
	GetUnread() int64
	New(level NotifyLevel, name string, message string, detail string) error
	GetAll(cond *Condition) ([]*Notification, error)
	Get(id uuid.UUID) (*Notification, error)
	Count(cond *Condition) (int64, error)
	Delete(id uuid.UUID) (bool, error)
	DeleteAll() (int64, error)
}

type SettingHandler interface {
	WithTx(tx *gorm.DB) SettingHandler
	BuildCache() error

	// GetAll return all settings.
	//
	// Copies are returned, modification will not be saved.
	GetAll() map[string]string

	// Get get name setting.
	Get(name string) (string, bool)

	// Set set setting name with value.
	Set(name string, value string) error

	// Delete delete name setting.
	Delete(name string) (bool, error)

	// DeleteAll delete all settings.
	DeleteAll() (int64, error)
}

File: skynet/src/handler.rs
use crate::{
    entity::{groups, notifications, permissions, users},
    permission::UserPerm,
    Condition, HyUuid, PermEntry, PermissionItem, Skynet,
};
use anyhow::Result;
use async_trait::async_trait;
use redis::aio::ConnectionManager;
use sea_orm::DatabaseTransaction;
use skynet_macro::default_handler;
use std::collections::HashMap;

/// User handler.
#[default_handler(users)]
#[async_trait]
pub trait UserHandler: Send + Sync {
    /// Create new user, generate random password if not set.
    /// Returned password is in plain text.
    async fn create(
        &self,
        db: &DatabaseTransaction,
        username: &str,
        password: Option<&str>,
        avatar: Option<Vec<u8>>,
    ) -> Result<users::Model>;

    /// Update user infos by `uid`.
    /// Password will be automatically hashed.
    ///
    /// User will be kicked if password is updated.
    async fn update(
        &self,
        db: &DatabaseTransaction,
        redis: &ConnectionManager,
        skynet: &Skynet,
        uid: &HyUuid,
        username: Option<&str>,
        password: Option<&str>,
        avatar: Option<Vec<u8>>,
    ) -> Result<users::Model>;

    /// Update login `uid` and `ip`.
    async fn update_login(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
        ip: &str,
    ) -> Result<users::Model>;

    /// Check `username` `password`.
    ///
    /// - If error, return `Err`.
    /// - If `username` not found, return `(false, None)`.
    /// - If `password` not match, return `(false, Some)`.
    /// - If success, return `(true, Some)`.
    async fn check_pass(
        &self,
        db: &DatabaseTransaction,
        username: &str,
        password: &str,
    ) -> Result<(bool, Option<users::Model>)>;

    /// Find user by `username`.
    /// Return `None` when not found.
    async fn find_by_name(
        &self,
        db: &DatabaseTransaction,
        username: &str,
    ) -> Result<Option<users::Model>>;

    /// Reset user password by `uid`.
    /// Return `None` when not found, password is in plain text otherwise.
    ///
    /// User will be kicked after reset.
    async fn reset(
        &self,
        db: &DatabaseTransaction,
        redis: &ConnectionManager,
        skynet: &Skynet,
        uid: &HyUuid,
    ) -> Result<Option<users::Model>>;

    /// Kick all `uid` login sessions.
    async fn kick(&self, redis: &ConnectionManager, skynet: &Skynet, uid: &HyUuid) -> Result<()>;

    /// Delete all users and kick them.
    async fn delete_all(
        &self,
        db: &DatabaseTransaction,
        redis: &ConnectionManager,
        skynet: &Skynet,
    ) -> Result<u64>;

    /// Delete `uid` users and kick them.
    async fn delete(
        &self,
        db: &DatabaseTransaction,
        redis: &ConnectionManager,
        skynet: &Skynet,
        uid: &[HyUuid],
    ) -> Result<u64>;
}

/// Group handler.
#[default_handler(groups)]
#[async_trait]
pub trait GroupHandler: Send + Sync {
    /// Link all `uid` user to all `gid` group.
    async fn link(&self, db: &DatabaseTransaction, uid: &[HyUuid], gid: &[HyUuid]) -> Result<()>;

    /// Unlinks user and group.
    ///
    /// - If `uid.is_empty()`, remove all users in each `gid`.
    /// - If `gid.is_empty()`, remove all groups in each `uid`.
    /// - Otherwise remove each `uid` with each `gid`.
    async fn unlink(&self, db: &DatabaseTransaction, uid: &[HyUuid], gid: &[HyUuid])
        -> Result<u64>;

    /// Update group infos by `gid`.
    async fn update(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        name: Option<&str>,
        note: Option<&str>,
    ) -> Result<groups::Model>;

    /// Create new user group.
    async fn create(
        &self,
        db: &DatabaseTransaction,
        name: &str,
        note: &str,
    ) -> Result<groups::Model>;

    /// Find group by `name`.
    async fn find_by_name(
        &self,
        db: &DatabaseTransaction,
        name: &str,
    ) -> Result<Option<groups::Model>>;

    /// Find group `gid` user by `uid`.
    async fn find_group_user_by_id(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        uid: &HyUuid,
    ) -> Result<Option<users::Model>>;

    /// Find group `gid` user.
    async fn find_group_user(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        cond: Condition,
    ) -> Result<(Vec<users::Model>, u64)>;

    /// Find user `uid` group.
    async fn find_user_group(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
        join: bool,
    ) -> Result<Vec<groups::Model>>;
}

/// Permission handler.
#[default_handler(permissions)]
#[async_trait]
pub trait PermHandler: Send + Sync {
    async fn find_or_init(
        &self,
        db: &DatabaseTransaction,
        name: &str,
        note: &str,
    ) -> Result<permissions::Model>;

    /// Find user `uid` permission.
    /// Note that user group permission is not included.
    async fn find_user(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
    ) -> Result<Vec<PermissionItem>>;

    /// Find group `gid` permission.
    async fn find_group(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
    ) -> Result<Vec<PermissionItem>>;

    /// Grant `uid` or `gid` with `pid` and `perm`.
    ///
    /// - If `uid` and `gid` both `None`, return `Ok` directly.
    /// - If `uid` and `gid` both `Some`, `pid` will be granted to all of them.
    async fn grant(
        &self,
        db: &DatabaseTransaction,
        uid: Option<&HyUuid>,
        gid: Option<&HyUuid>,
        pid: &HyUuid,
        perm: UserPerm,
    ) -> Result<()>;

    /// Create new user `uid` permission `perm`.
    async fn create_user(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
        perm: &[PermEntry],
    ) -> Result<()>;

    /// Create new group `gid` permission `perm`.
    async fn create_group(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        perm: &[PermEntry],
    ) -> Result<()>;

    /// Delete user `uid` permission `pid`.
    ///
    /// If `pid` is `None`, all user permission will be deleted.
    async fn delete_user(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
        pid: Option<&HyUuid>,
    ) -> Result<u64>;

    /// Delete group `gid` permission `pid`.
    ///
    /// If `pid` is `None`, all group permission will be deleted.
    async fn delete_group(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        pid: Option<&HyUuid>,
    ) -> Result<u64>;
}

/// Notification handler.
#[default_handler(notifications)]
#[async_trait]
pub trait NotificationHandler: Send + Sync {}

#[async_trait]
pub trait SettingHandler: Send + Sync {
    /// Build setting cache.
    async fn build_cache(&self, db: &DatabaseTransaction) -> Result<()>;

    /// Get all settings.
    fn get_all(&self) -> HashMap<String, String>;

    /// Get `name` setting.
    fn get(&self, name: &str) -> Option<String>;

    /// Set setting `name` with `value`.
    async fn set(&self, db: &DatabaseTransaction, name: &str, value: &str) -> Result<()>;

    /// Delete `name` setting.
    async fn delete(&self, db: &DatabaseTransaction, name: &str) -> Result<bool>;

    /// Delete all settings.
    async fn delete_all(&self, db: &DatabaseTransaction) -> Result<u64>;
}
